src/wp-includes/abilities-api/class-wp-ability.php

```diff
@@ -0,0 +1,618 @@
+	 *                                                       Receives optional mixed input and returns bool or WP_Error.
+	 *     @type array<string, mixed> $input_schema          Optional. JSON Schema definition for the ability's input.
+	 *     @type array<string, mixed> $output_schema         Optional. JSON Schema definition for the ability's output.
+	 *     @type array<string, mixed> $meta                  {
+	 *         Optional. Additional metadata for the ability.
+	 *
+	 *         @type array<string, null|bool> $annotations  Optional. Annotation metadata for the ability.
+	 *         @type bool                     $show_in_rest Optional. Whether to expose this ability in the REST API. Default false.
+	 *     }
+	 * }
+	 */
+	public function __construct( string $name, array $args ) {
+		$this->name = $name;
+
+		$properties = $this->prepare_properties( $args );
+
+		foreach ( $properties as $property_name => $property_value ) {
+			if ( ! property_exists( $this, $property_name ) ) {
+				_doing_it_wrong(
+					__METHOD__,
+					sprintf(
+						/* translators: %s: Property name. */
+						__( 'Property "%1$s" is not a valid property for ability "%2$s". Please check the %3$s class for allowed properties.' ),
+						'<code>' . esc_html( $property_name ) . '</code>',
+						'<code>' . esc_html( $this->name ) . '</code>',
+						'<code>' . self::class . '</code>'
+					),
+					'6.9.0'
+				);
+				continue;
+			}
+
+			$this->$property_name = $property_value;
+		}
+	}
+
+	/**
+	 * Prepares and validates the properties used to instantiate the ability.
+	 *
+	 * Errors are thrown as exceptions instead of WP_Errors to allow for simpler handling and overloading. They are then
+	 * caught and converted to a WP_Error when by WP_Abilities_Registry::register().
+	 *
+	 * @since 6.9.0
+	 *
+	 * @see WP_Abilities_Registry::register()
+	 *
+	 * @param array<string, mixed> $args {
+	 *     An associative array of arguments used to instantiate the ability class.
+	 *
+	 *     @type string               $label                 The human-readable label for the ability.
+	 *     @type string               $description           A detailed description of what the ability does.
+	 *     @type string               $category              The ability category slug this ability belongs to.
+	 *     @type callable             $execute_callback      A callback function to execute when the ability is invoked.
+	 *                                                       Receives optional mixed input and returns mixed result or WP_Error.
+	 *     @type callable             $permission_callback   A callback function to check permissions before execution.
+	 *                                                       Receives optional mixed input and returns bool or WP_Error.
+	 *     @type array<string, mixed> $input_schema          Optional. JSON Schema definition for the ability's input.
```

Comment:
nit: `$input_schema` is only optional when there are no inputs, otherwise it's required:

https://github.com/WordPress/wordpress-develop/blob/d3e1cd72a9191426de510a8f225d898a07575e01/src/wp-includes/abilities-api/class-wp-ability.php#L413-L426


```suggestion
	 *     @type array<string, mixed> $input_schema          Optional. JSON Schema definition for the ability's input. Required if ability accepts an input.
```